nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/service/ColdStorageServiceImpl.java

```
@@ -326,7 +330,25 @@ public DocumentModel retrieveFromColdStorage(CoreSession session, DocumentRef do
         BlobStatus blobStatus = ColdStorageHelper.getBlobStatus(documentModel);
         Function<DocumentModel, Boolean> doNotify;
         DocumentModel docResult = null;
+        Blob coldContent = (Blob) documentModel.getPropertyValue(COLD_STORAGE_CONTENT_PROPERTY);
+        String key = getContentBlobKey(coldContent);
         if (ColdStorageHelper.isDownloadable(blobStatus)) {
+            Instant downloadableUntil = blobStatus.getDownloadableUntil();
+            if (downloadableUntil == null) {
+                log.warn(
+                        "Cold content blob: {} of document: {} appears to be already restored. Restoring the Nuxeo document.",
+                        key, documentRef);
+                // Blob is restored but the doc is not, let's restore the doc to have a consistent state
+                proceedRestoreMainContent(session, documentModel, false, false);
+                // Fire event for audit purpose
+                fireEvent(documentModel, documentModel.getCoreSession(), COLD_STORAGE_CONTENT_TO_RESTORE_EVENT_NAME);
+                // Throw an error because we are doing a restore instead of retrieve
+                // But let's commit before to avoid a rollback
+                TransactionHelper.commitOrRollbackTransaction();
```

The transaction should not be managed at this level. 
If the user restores (instead of retrieve) the doc is it going to work ? 
if yes we should throw an exception asking to restore the doc in order to fix it.

---

The pbm is that the doc is in state "sent to cold storage" but the blob in the s3 bucket is restored. We want to fix this inconsistent state by making the nuxeo doc in-state restored as well.

> If the user restores (instead of retrieve) the doc is it going to work ?

yes

> if yes we should throw an exception asking to restore the doc in order to fix it.

sounds like this happens during propagation, I don't think throwing an error asking to restore the doc would lead anyone to do so...

---

I am ok to restore the doc seamlessly and do not throw error and do not manage the transaction such a way